### PR TITLE
ConEmu Console Emulator Control for Running Git Commands in the Real Terminal

### PR DESCRIPTION
**Pull request by hypersw**
Terminal window embedded in the commands form:

untitled20160316221556

This also fixes interactivity with OpenSSH prompts:

untitled20160316221905

And even allows to run interactive rebase in `vi` right in this window, if you'd like.

untitled20160316222121

Bonus: a terminal in a tab on the repo view:

untitled20160316222303

There is an option to switch back to the editbox control (which actually happens if you're not running on Windows NT). So the control is now a switchable implementation.

### CODE REVIEW

**Comment by hypersw** [comment]
A build to try:
https://github.com/JetBrains/gitextensions/releases/tag/2016-03-14

**Comment by hypersw** [comment]
Fixes https://github.com/gitextensions/gitextensions/issues/1393.

**Comment by EbenZhang** [comment]
Can you normailse the tabs in the designer.cs to be spaces, it quite annoying when solving conflicts?

**Comment by EbenZhang** [comment]
For `bash` and `gitbash`, please append the  ` --login -i` as the argument, otherwise `.bash_profile` wouldn't work.

File: GitUI/CommandsDialogs/FormBrowse.cs [hunk 1]
```diff
@@ -3352,13 +3357,64 @@ private void GitTreeKeyDown(object sender, KeyEventArgs e)
             if (e.KeyCode == Keys.Enter || e.KeyCode == Keys.Return)
             {
                 if (GitTree.SelectedNode != null)
-                {
+                {
```
Comment (IDisposable) [inline GitUI/CommandsDialogs/FormBrowse.cs]: excess blank

File: GitUI/CommandsDialogs/FormBrowse.cs [hunk 2]
```diff
@@ -3352,13 +3357,64 @@ private void GitTreeKeyDown(object sender, KeyEventArgs e)
...
+			    if(terminal.IsConsoleEmulatorOpen) // If user has typed "exit" in there, restart the shell; otherwise just return
+				    return;
+
+			    // Create the terminal
+			    var startinfo = new ConEmuStartInfo();
+			    startinfo.StartupDirectory = Module.WorkingDir;
+			    startinfo.WhenConsoleProcessExits = WhenConsoleProcessExits.CloseConsoleEmulator;
+
+			    // Choose the console: bash from git with fallback to cmd
+			    string sGitBashFromUsrBin = "";/*This is not a console program and is not reliable yet, suppress for now.*/ //Path.Combine(Path.Combine(Path.Combine(AppSettings.GitBinDir, ".."), ".."), "git-bash.exe"); // Git bin dir is /usr/bin under git installdir, so go 2x up
```
Comment (EbenZhang) [inline GitUI/CommandsDialogs/FormBrowse.cs]: `GitBinDir` is for linux... Should it be `GitDir`?

**Comment by EbenZhang** [comment]
The `force push` I mentioned before is because this call `form.Retry();`  in `FormPush.cs` the `Retry()` will start another console process but there's already one running, thus results in an exception in `processStart` method in `FormProcess.cs`.

I'm not sure how to fix it. Maybe close the process or reuse the process?

**Comment by EbenZhang** [comment]
The cursor shows as a rectangle, any idea to fix it? 
If type exit and reopen the tab, the cursor shows as normal.

**Comment by EbenZhang** [comment]
In the form process, it seems the conemu always scrolls to the second line.
conemu_scrollissue

**Comment by ghanique** [comment]
I'd live this feature! The ssh-pushing regularly fails for no specific
reason and I can't fix it - being able to type "yes" or to type a password
would be great.

BTW: does anyone know where this feature comes from? Or what issue it
solves? (I'm loving it and hope it will make it to the next release).
[image: Inline image 1]

**Comment by Maximus5** [comment]
Cursor is a "Rectangle" if ConEmu has no focus, e.g. focus is on "OK" key. It's configurable: https://conemu.github.io/en/SettingsTextCursor.html

First line went out because there are more lines in the output than fit in visible area.

**Comment by EbenZhang** [comment]
> First line went out because there are more lines in the output than fit in visible area.

happens even if 2 lines

**Comment by EbenZhang** [comment]
> Cursor is a "Rectangle" if ConEmu has no focus, e.g. focus is on "OK" key. It's configurable:

when it shows rectangle, i can still type, so it is focused

**Comment by hypersw** [comment]
@Maximus5: with the main page tab, there's a bug with the ConEmu control in *some* cases when it fails to detect focus on the first show. It would not repro on all the machines.

**Comment by Maximus5** [comment]
To fix problem with focus detection I need [LogFiles](https://conemu.github.io/en/LogFiles.html).

**Comment by EbenZhang** [comment]
> To fix problem with focus detection I need LogFiles.

That is for the exe. Any way to specify the log argument for the ConEmu.Winforms.dll

**Comment by hypersw** [comment]
* Force push: OK I'll check that; with fresh ConEmu Control I've changed the policy to respawn the process rather than complain, so it should have gotten better.

* Scrolling one line: have not observed this, and we've been running early versions for all 2016 (to address the openssh interactivity issues). Could you check if this has anything to do to the form size you've got stored in the settings? If you resize it to smth bigger and it remembers the size, would this happen?

**Comment by EbenZhang** [comment]
> Could you check if this has anything to do to the form size you've got stored in the settings? If you resize it to smth bigger and it remembers the size, would this happen?

Yes, it's the size causes the problem. Solved if I make it a little bit bigger. Thanks.

**Comment by hypersw** [comment]
I think it's the sequence of "creating form in the default size => laid out ConEmu control for this size => ConEmu writes text to stdout => MRU size settings are applied, window downscaled => ConEmu sized down, makes a scroll". Will check.

**Comment by hypersw** [comment]
I'll try to play with `GitExtensionsForm::RestorePosition`, it also got issues with secondary windows falling offscreen when e.g. working with submodules on smth like `git checkout` for gitext. But I'd rather not entangle this work with the current PR.

**Comment by EbenZhang** [comment]
@hypersw  @KindDragon  The force push is still a problem.

**Comment by hypersw** [comment]
@EbenZhang: is there an issue for that?
I'll do it sooner now that it's merged; before that it were unclear if it were getting into master, and for our internal uses that were quite a low priority.

**Comment by karyon** [comment]
additionally, i also noticed what @EbenZhang showed here https://github.com/gitextensions/gitextensions/pull/3152#issuecomment-197876808, that it always skips a line.

**Comment by Maximus5** [comment]
In the current ConEmu build you may enable LogFiles on the "Features" settings page. LogFiles may help to identify the problem with skipped line.

**Comment by hypersw** [comment]
@Maximus5: this can be easily reproduced if you make the window narrower, it scrolls out the single line. I'll try to model this in the conemu-inside repo samples.

**Comment by EbenZhang** [comment]
> @EbenZhang: is there an issue for that?
> I'll do it sooner now that it's merged; before that it were unclear if it were getting into master, and for our internal uses that were quite a low priority.

Created one #3173

**Comment by jazzzz** [comment]
Is there something to do to enable ConEmu on master? I build it with `BuildInstallers.VS2015.cmd` and ConEmu is not enabled.

File: GitUI/CommandsDialogs/SettingsDialog/Pages/AdvancedSettingsPage.cs [hunk 3]
```diff
@@ -19,6 +19,7 @@ protected override void SettingsToPage()
             chkAlwaysShowAdvOpt.Checked = AppSettings.AlwaysShowAdvOpt;
             chkCheckForRCVersions.Checked = AppSettings.CheckForReleaseCandidates;
             chkRememberIgnoreWhiteSpacePreference.Checked = AppSettings.RememberIgnoreWhiteSpacePreference;
+            chkConsoleEmulator.Checked = AppSettings.UseConsoleEmulatorForCommands;
```
Comment (jbialobr) [inline GitUI/CommandsDialogs/SettingsDialog/Pages/AdvancedSettingsPage.cs]: This setting went under the Advanced settings page. Which of the actions: disabling/enabling this setting is considered as an advanced one and why?

**Pull request merged**